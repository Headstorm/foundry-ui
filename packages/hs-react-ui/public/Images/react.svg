<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 19.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_2" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 width="561.8px" height="499.9px" viewBox="-80.4 151.2 561.8 499.9" style="enable-background:new -80.4 151.2 561.8 499.9;"
	 xml:space="preserve">
<style type="text/css">
	.st0{fill:#939598;}
</style>
<circle class="st0" cx="200.5" cy="400.6" r="50.2"/>
<path class="st0" d="M200.5,515.6c-70.5,0-132.1-8.3-178.2-24.1c-29.9-10.2-55.3-23.8-73.4-39.3c-19.2-16.4-29.4-34.3-29.4-51.6
	c0-33.2,36.4-65.7,97.5-86.9c50-17.4,115.2-27.1,183.4-27.1c67,0,131.3,9.4,181,26.6c29.1,10,53.6,23,71,37.4
	c18.9,15.8,28.9,33.1,28.9,50c0,34.5-40.7,69.4-106.3,91.1C328.6,507.1,266.6,515.6,200.5,515.6z M200.5,310.6
	c-64.7,0-128.7,9.4-175.5,25.7c-40.5,14.1-81.4,39.5-81.4,64.3c0,10,7.4,21.8,21,33.4c15.8,13.5,38.4,25.6,65.5,34.8
	c43.6,14.9,102.6,22.8,170.4,22.8c63.6,0,122.9-8,167-22.7c61.7-20.5,89.9-49.8,89.9-68.3c0-9.5-7.2-20.7-20.3-31.6
	c-15.1-12.6-37.1-24.1-63.4-33.2C326.4,319.6,264.9,310.6,200.5,310.6z"/>
<path class="st0" d="M86.6,650.8c-10.2,0-19.2-2.2-26.8-6.6C31,627.6,21.1,579.8,33.1,516.3c9.9-52.1,34.1-113.3,68.2-172.4h0
	c33.5-58,73.7-109,113.4-143.5c23.2-20.2,46.7-35,67.9-42.8c23.1-8.5,43.1-8.5,57.7-0.1c29.9,17.2,39.8,70,25.8,137.6
	c-9.9,48-33.5,105.9-66.5,163.2c-35.2,61-73.2,110.2-109.9,142.3c-23.8,20.8-48.3,36-70.7,43.9C107.4,648.7,96.4,650.8,86.6,650.8z
	 M122.1,355.9c-32.3,56-56.2,116.1-65.4,164.9c-8,42.1-6.5,90.2,15,102.6c16.1,9.3,55,0.5,102.3-40.9c34.7-30.3,71-77.5,104.9-136.2
	c31.8-55.1,54.4-110.5,63.8-156c6-29.1,7.3-55.5,3.8-76.4c-3-17.9-9.4-30.6-18.2-35.6c-8.2-4.7-21.5-4.1-37.5,1.8
	c-18.5,6.8-39.4,20.1-60.4,38.4C192.8,251.3,154.3,300.1,122.1,355.9L122.1,355.9z"/>
<path class="st0" d="M314.4,651.1c-27.2,0-61.7-16.4-97.7-47.4c-40.2-34.6-81.1-86.1-115.3-145.2v0c-33.6-58-57.6-118.3-67.7-170
	c-5.9-30.2-7-57.9-3.2-80.2c4.2-24.3,14.1-41.6,28.8-50.1c29.8-17.3,80.5,0.5,132.1,46.4c36.6,32.5,75,81.9,108.1,139.1
	c35.3,61,59,118.5,68.4,166.3c6.1,31,7.1,59.8,2.8,83.2c-4.6,24.9-15,42.6-30,51.3C333.2,648.9,324.3,651.1,314.4,651.1z
	 M122.2,446.5c32.4,56,72.6,106.7,110.2,139c32.5,28,74.9,50.6,96.4,38.2c16.1-9.3,27.9-47.4,15.7-109
	c-9-45.2-31.7-100.2-65.6-158.9c-31.9-55.1-68.6-102.4-103.3-133.2c-22.2-19.7-44.4-34.1-64.2-41.5c-17-6.4-31.2-7.1-39.9-2.1
	c-8.2,4.7-14.3,16.6-17.2,33.4c-3.3,19.4-2.3,44.2,3.1,71.5C66.9,333,89.9,390.7,122.2,446.5L122.2,446.5z"/>
</svg>
