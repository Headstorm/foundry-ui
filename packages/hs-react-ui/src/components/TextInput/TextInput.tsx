import React, { ChangeEvent, EventHandler, ReactNode, SyntheticEvent, useCallback } from 'react';
import styled, { css, StyledComponentBase } from 'styled-components';
import Icon from '@mdi/react';
import { mdiClose } from '@mdi/js';
import debounce from 'lodash.debounce';
import { Div, Input as InputElement, TextArea } from '../../htmlElements';
import { SubcomponentPropsType } from '../commonTypes';
import { useTheme } from '../../context';
import { disabledStyles } from '../../utils/color';

const Container = styled(Div)`
  ${({ disabled = false, isValid }: { disabled?: boolean; isValid?: boolean }) => {
    const { colors } = useTheme();
    return `
      border 2px solid ${isValid === false ? colors.destructive : colors.grayMedium};
      min-width: 10rem;
      position: relative;
      display: flex;
      flex-flow: row;
      border-radius: 0.25em;
      background-color: ${colors.background};
      ${disabled ? disabledStyles() : ''}
  `;
  }}
`;

const TextInputContainer = styled(InputElement)`
  ${() => {
    const { colors } = useTheme();
    return `
      border: 0 none;
      flex-grow: 1;
      outline: 0 none;
      height: 2em;
      font-size: 1em;
      padding: 0.5rem;
      background-color: ${colors.transparent};
  `;
  }}
`;

const TextAreaInputContainer = styled(TextArea)`
  ${({ multiLineIsResizable }: TextInputProps) => {
    const { colors } = useTheme();
    return `
      border: 0 none;
      flex-grow: 1;
      outline: 0 none;
      font-size: 1em;
      min-height: 2em;
      min-width: 0px;
      padding: .5rem;
      background-color: ${colors.transparent};
      resize: ${multiLineIsResizable ? 'both' : 'none'};
    `;
  }}
`;

const IconContainer = styled(Div)`
  ${() => {
    const { colors } = useTheme();

    return `
      padding: 0.5em;
      height: 100%;
      display: flex;
      align-items: center;
      justify-content: center;
      color: ${colors.grayMedium};
      cursor: pointer;
    `;
  }}
`;

const CharacterCounter = styled(Div)`
  ${({ textIsTooLong, isValid, errorMessage }) => {
    const { colors } = useTheme();
    return css`
      position: absolute;
      top: calc(100% + ${isValid && (!errorMessage || errorMessage !== '') ? '0.25em' : '2em'});
      right: 0.25em;
      color: ${textIsTooLong ? colors.destructive : colors.grayLight};
    `;
  }}
`;

const ErrorContainer = styled(Div)`
  ${() => {
    const { colors } = useTheme();
    return css`
      position: absolute;
      top: calc(100% + 0.25em);
      color: ${colors.destructive};
      font-size: 0.75rem;
    `;
  }}
`;

export type TextInputProps = InputHTMLAttributes<HTMLInputElement> & TextareaHTMLAttributes<HTMLTextAreaElement> & {
  iconPrefix?: string | ReactNode;
  onClear?: (event: SyntheticEvent) => void;
  onChange?: EventHandler<ChangeEvent<HTMLInputElement>>;
  debouncedOnChange?: EventHandler<ChangeEvent<HTMLInputElement>>;
  onKeyPress?: (event: SyntheticEvent) => void;
  onKeyDown?: (event: SyntheticEvent) => void;
  onKeyUp?: (event: SyntheticEvent) => void;
  onBlur?: (event: SyntheticEvent) => void;
  onInput?: (event: SyntheticEvent) => void;
  onFocus?: (event: SyntheticEvent) => void;
  onReset?: (event: SyntheticEvent) => void;
  cols?: number;
  rows?: number;
  value?: string;
  disabled?: boolean;
  defaultValue?: string;
  isValid?: boolean;
  isMultiline?: boolean;
  errorMessage?: string;
  debounceInterval?: number;
  multiLineIsResizable?: boolean;
  maxLength?: number;
  allowTextBeyondMaxLength?: boolean;
  showCharacterCount?: boolean;

  StyledContainer?: string & StyledComponentBase<any, {}>;
  StyledInput?: string & StyledComponentBase<any, {}>;
  StyledIconContainer?: string & StyledComponentBase<any, {}>;
  StyledErrorContainer?: string & StyledComponentBase<any, {}>;
  containerProps?: SubcomponentPropsType;
  inputProps?: SubcomponentPropsType;
  iconContainerProps?: SubcomponentPropsType;
  errorContainerProps?: SubcomponentPropsType;
};

const createIcon = (
  StyledIconContainer: string & StyledComponentBase<any, {}>,
  iconPrefix: ReactNode,
) => {
  if (typeof iconPrefix === 'string') {
    return (
      <StyledIconContainer>
        <Icon size="1rem" path={iconPrefix} />
      </StyledIconContainer>
    );
  }

  return <StyledIconContainer>{iconPrefix}</StyledIconContainer>;
};

const defaultCallback = () => {}; // eslint-disable-line @typescript-eslint/no-empty-function

const TextInput = ({
  debouncedOnChange = defaultCallback,
  onKeyPress,
  onKeyDown,
  onKeyUp,
  onBlur,
  onInput,
  onFocus,
  onReset,
  cols = 10,
  rows = 10,
  value,
  defaultValue = '',
  isValid = true,
  isMultiline,
  errorMessage,
  ariaLabel,
  type = 'text',
  disabled = false,
  debounceInterval = 8,
  multiLineIsResizable,
  maxLength,
  allowTextBeyondMaxLength = false,
  showCharacterCount = false,

  StyledContainer = Container,
  StyledIconContainer = IconContainer,
  StyledErrorContainer = ErrorContainer,
  containerProps = {},
  inputProps = {},
  iconContainerProps = {},
  errorContainerProps = {},
}: TextInputProps) => {
  // Debounce the change function using useCallback so that the function is not initialized each time it renders
  const debouncedChange = useCallback(debounce(debouncedOnChange, debounceInterval), [
    debouncedOnChange,
    debounceInterval,
  ]);

  // Determine the correct input type. Using a single input and the 'as' keyword
  // to display as a text area disables the ability to set cols/rows
  let InputComponent: string & StyledComponentBase<any, {}> = TextInputContainer;
  if (StyledInput) {
    InputComponent = StyledInput;
  } else if (isMultiline) {
    InputComponent = TextAreaInputContainer;
  }
  const displayValue = value || defaultValue;

  return (
    <StyledContainer disabled={disabled} isValid={isValid} {...containerProps}>
      {iconPrefix && createIcon(StyledIconContainer, iconPrefix)}
      {/*
        // @ts-ignore */}
      <InputComponent
        cols={cols}
        rows={rows}
        aria-label={ariaLabel}
        placeholder={placeholder}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
          e.persist();
          if (maxLength && maxLength >= 0) {
            e.target.value = allowTextBeyondMaxLength
              ? e.target.value
              : e.target.value.slice(0, maxLength);
          }
          onChange(e);
          debouncedChange(e);
        }}
        onKeyPress={onKeyPress}
        onKeyDown={onKeyDown}
        onKeyUp={onKeyUp}
        onFocus={onFocus}
        onBlur={onBlur}
        onReset={onReset}
        onInput={onInput}
        value={displayValue}
        id={id}
        type={type}
        multiLineIsResizable={multiLineIsResizable}
        {...inputProps}
      />
      {onClear && (
        <StyledIconContainer onClick={onClear} {...iconContainerProps}>
          <Icon path={mdiClose} size="1em" />
        </StyledIconContainer>
      )}
      {showCharacterCount && maxLength && (
        <CharacterCounter
          errorMessage={errorMessage}
          isValid={isValid}
          textIsTooLong={displayValue.length > maxLength}
        >
          {displayValue.length} / {maxLength}
        </CharacterCounter>
      )}
      {isValid === false && errorMessage && (
        <StyledErrorContainer {...errorContainerProps}>{errorMessage}</StyledErrorContainer>
      )}
    </StyledContainer>
  );
};

TextInput.Container = Container;
TextInput.ErrorContainer = ErrorContainer;
TextInput.Input = TextInputContainer;
TextInput.IconContainer = IconContainer;
TextInput.TextArea = TextAreaInputContainer;

export default TextInput;
