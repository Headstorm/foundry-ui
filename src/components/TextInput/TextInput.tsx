import React, {
  ChangeEvent,
  EventHandler,
  ReactNode,
  SyntheticEvent,
  useCallback,
  useState,
  TextareaHTMLAttributes,
  InputHTMLAttributes,
  useRef,
} from 'react';
import styled, { css } from 'styled-components';
import Icon from '@mdi/react';
import { mdiClose } from '@mdi/js';
import debounce from 'lodash.debounce';
import { mergeRefs } from '../../utils/refs';
import { Div, Input as InputElement, TextArea } from '../../htmlElements';
import { SubcomponentPropsType, StyledSubcomponentType } from '../commonTypes';
import { useAnalytics, useTheme } from '../../context';
import { disabledStyles } from '../../utils/color';

const Container = styled(Div)`
  ${({ disabled = false, isValid }: { disabled?: boolean; isValid?: boolean }) => {
    const { colors } = useTheme();
    return `
      border 1px solid ${isValid === false ? colors.destructive : colors.grayMedium};
      min-width: 10rem;
      position: relative;
      display: flex;
      flex-flow: row;
      border-radius: 0.25em;
      background-color: ${colors.background};
      ${disabled ? disabledStyles() : ''}
  `;
  }}
`;

const TextInputContainer = styled(InputElement)`
  ${() => {
    const { colors } = useTheme();
    return `
      border: 0 none;
      flex-grow: 1;
      outline: 0 none;
      font-size: 1em;
      padding: 0.5rem;
      background-color: ${colors.transparent};
      &:focus {
        outline: none;
        box-shadow: 0 0 5px 0.150rem ${colors.tertiary};
  `;
  }}
`;

const TextAreaInputContainer = styled(TextArea)`
  ${({ multiLineIsResizable }: TextInputProps) => {
    const { colors } = useTheme();
    return `
      border: 0 none;
      flex-grow: 1;
      outline: 0 none;
      font-size: 1em;
      min-width: 0px;
      padding: .5rem;
      background-color: ${colors.transparent};
      resize: ${multiLineIsResizable ? 'both' : 'none'};
      &:focus {
        outline: none;
        box-shadow: 0 0 5px 0.150rem ${colors.tertiary};
    `;
  }}
`;

const IconContainer = styled(Div)`
  ${() => {
    const { colors } = useTheme();
    return `
      padding: 0.5em;
      height: 100%;
      align-items: center;
      justify-content: center;
      color: ${colors.grayMedium};
      cursor: pointer;
    `;
  }}
`;

const CharacterCount = styled(Div)`
  ${({ textIsTooLong, isValid, errorMessage }) => {
    const { colors } = useTheme();
    return css`
      position: absolute;
      top: calc(100% + ${isValid && (!errorMessage || errorMessage !== '') ? '0.25em' : '2em'});
      right: 0.25em;
      color: ${textIsTooLong ? colors.destructive : colors.grayLight};
    `;
  }}
`;

const ErrorContainer = styled(Div)`
  ${() => {
    const { colors } = useTheme();
    return css`
      position: absolute;
      top: calc(100% + 0.25em);
      color: ${colors.destructive};
      font-size: 0.75rem;
    `;
  }}
`;

export type TextInputProps = InputHTMLAttributes<HTMLInputElement> &
  TextareaHTMLAttributes<HTMLTextAreaElement> & {
    iconPrefix?: string | ReactNode;
    onClear?: (event: SyntheticEvent) => void;
    clearable?: boolean;
    debouncedOnChange?: EventHandler<ChangeEvent<HTMLInputElement>>;
    isValid?: boolean;
    isMultiline?: boolean;
    errorMessage?: string;
    debounceInterval?: number;
    multiLineIsResizable?: boolean;
    maxLength?: number;
    allowTextBeyondMaxLength?: boolean;
    showCharacterCount?: boolean;

    StyledContainer?: StyledSubcomponentType;
    StyledInput?: StyledSubcomponentType;
    StyledIconContainer?: StyledSubcomponentType;
    StyledErrorContainer?: StyledSubcomponentType;
    StyledTextArea?: StyledSubcomponentType;
    StyledCharacterCount?: StyledSubcomponentType;

    containerProps?: SubcomponentPropsType;
    inputProps?: SubcomponentPropsType;
    iconContainerProps?: SubcomponentPropsType;
    errorContainerProps?: SubcomponentPropsType;
    characterCountProps?: SubcomponentPropsType;

    containerRef?: React.RefObject<HTMLDivElement>;
    inputRef?: React.RefObject<HTMLInputElement>;
    errorContainerRef?: React.RefObject<HTMLDivElement>;
    characterCountRef?: React.RefObject<HTMLDivElement>;
  };

const createIcon = (StyledIconContainer: StyledSubcomponentType, iconPrefix: ReactNode) => {
  if (typeof iconPrefix === 'string') {
    return (
      <StyledIconContainer>
        <Icon size="1rem" path={iconPrefix} />
      </StyledIconContainer>
    );
  }

  return <StyledIconContainer>{iconPrefix}</StyledIconContainer>;
};

const defaultCallback = () => {}; // eslint-disable-line @typescript-eslint/no-empty-function

const TextInput = ({
  debouncedOnChange = defaultCallback,
  onClear,
  clearable,
  iconPrefix,
  isValid = true,
  isMultiline,
  errorMessage,
  debounceInterval = 8,
  multiLineIsResizable,
  maxLength,
  allowTextBeyondMaxLength = false,
  showCharacterCount = false,

  StyledContainer = Container,
  StyledInput = TextInputContainer,
  StyledIconContainer = IconContainer,
  StyledErrorContainer = ErrorContainer,
  StyledTextArea = TextAreaInputContainer,
  StyledCharacterCount = CharacterCount,

  containerProps = {},
  inputProps = {},
  iconContainerProps = {},
  errorContainerProps = {},
  characterCountProps = {},

  containerRef,
  inputRef,
  errorContainerRef,
  characterCountRef,

  ...nativeHTMLAttributes
}: TextInputProps): JSX.Element => {
  const handleEventWithAnalytics = useAnalytics();

  const handleDebouncedOnChange = (e: any) =>
    handleEventWithAnalytics(
      'TextInput',
      debouncedOnChange,
      'debouncedOnChange',
      e,
      containerProps,
    );

  // Debounce the change function using useCallback so that the function is not initialized each time it renders
  const debouncedChange = useCallback(debounce(handleDebouncedOnChange, debounceInterval), [
    handleDebouncedOnChange,
    debounceInterval,
  ]);

  const internalInputRef = useRef<HTMLInputElement>(null);

  const InputComponent: StyledSubcomponentType = isMultiline ? StyledTextArea : StyledInput;

  const [internalValue, setInternalValue] = useState(
    nativeHTMLAttributes.value || nativeHTMLAttributes.defaultValue || '',
  );

  const handleClear = (e: any) => {
    const onClearToUse =
      onClear &&
      (() => {
        if (!inputRef?.current?.value && internalInputRef?.current?.value) {
          setInternalValue('');
          internalInputRef.current.value = '';
        }
      });
    handleEventWithAnalytics('TextInput', onClearToUse, 'onClear', e, containerProps);
  };

  if (internalValue !== nativeHTMLAttributes.value && nativeHTMLAttributes.value === '') {
    setInternalValue('');
  }

  return (
    <StyledContainer
      disabled={nativeHTMLAttributes.disabled}
      isValid={isValid}
      ref={containerRef}
      {...containerProps}
    >
      {iconPrefix && createIcon(StyledIconContainer, iconPrefix)}
      <InputComponent
        // Set default values above nativeHTMLAttributes
        type="text"
        disabled={false}
        cols={10}
        rows={10}
        {...nativeHTMLAttributes}
        onChange={(e: ChangeEvent<HTMLInputElement>) => {
          e.persist();
          if (maxLength && maxLength >= 0) {
            e.target.value = allowTextBeyondMaxLength
              ? e.target.value
              : e.target.value.slice(0, maxLength);
          }
          setInternalValue(e.target.value);
          if (nativeHTMLAttributes.onChange) {
            nativeHTMLAttributes.onChange(e);
          }
          debouncedChange(e);
        }}
        multiLineIsResizable={multiLineIsResizable}
        ref={mergeRefs([inputRef, internalInputRef])}
        {...inputProps}
      />
      {(onClear || clearable) && (
        <StyledIconContainer onClick={handleClear} {...iconContainerProps}>
          <Icon path={mdiClose} size="1em" />
        </StyledIconContainer>
      )}
      {showCharacterCount && (
        <StyledCharacterCount
          ref={characterCountRef}
          errorMessage={errorMessage}
          isValid={isValid}
          textIsTooLong={maxLength ? (internalValue as string).length > maxLength : false}
          {...characterCountProps}
        >
          {(internalValue as string).length} {maxLength !== undefined ? `/ ${maxLength}` : null}
        </StyledCharacterCount>
      )}
      {isValid === false && errorMessage && (
        <StyledErrorContainer ref={errorContainerRef} {...errorContainerProps}>
          {errorMessage}
        </StyledErrorContainer>
      )}
    </StyledContainer>
  );
};

TextInput.Container = Container;
TextInput.ErrorContainer = ErrorContainer;
TextInput.Input = TextInputContainer;
TextInput.IconContainer = IconContainer;
TextInput.TextArea = TextAreaInputContainer;
TextInput.CharacterCount = CharacterCount;

export default TextInput;
