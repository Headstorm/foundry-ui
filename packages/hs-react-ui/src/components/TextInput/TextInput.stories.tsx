import React, { useState, SyntheticEvent } from 'react';
import { text, select, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import Icon from '@mdi/react';
import * as IconPaths from '@mdi/js';
import { storiesOf, addDecorator } from '@storybook/react';
import { withA11y } from '@storybook/addon-a11y';
import { withDesign } from 'storybook-addon-designs';
import TextInput from './TextInput';

addDecorator(withA11y);
addDecorator(withDesign);

const design = {
  type: 'figma',
  url: 'https://www.figma.com/file/3r2G00brulOwr9j7F6JF59/Generic-UI-Style?node-id=102%3A29',
};

storiesOf('TextInput', module)
  .add(
    'Basic Text Input',
    () => {
      const [inputValue, setInputValue] = useState('');
      const options = {
        none: '',
        ...IconPaths,
      };
      const getIconPath = (path: string) => (path ? <Icon size="16px" path={path} /> : null);

      return (
        <TextInput
          onChange={(event: SyntheticEvent) => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            const newValue = event.target.value;
            setInputValue(newValue);
            action('onChange')(newValue);
          }}
          value={inputValue}
          placeholder={text('placeholder', 'Place Holder')}
          onClear={
            boolean('onClear?', false)
              ? () => {
                  setInputValue('');
                  action('onClear')();
                }
              : undefined
          }
          iconPrefix={getIconPath(select('iconPrefix', options, options.mdiComment))}
          isMultiline={boolean('isMultiline?', false)}
          errorMessage={text('errorMessage', '')}
        />
      );
    },
    { design },
  )
  .add(
    'Text Input with all the gadgets',
    () => {
      const [inputValue, setInputValue] = useState('');
      const options = {
        none: '',
        ...IconPaths,
      };
      const getIconPath = (path: string) => (path ? <Icon size="16px" path={path} /> : null);
      const isMultiline = inputValue.length > 15;
      const isError =
        inputValue.length < 5 && inputValue.length > 0 ? 'Short Message Error' : undefined;

      return (
        <TextInput
          onChange={(event: SyntheticEvent) => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            const newValue = event.target.value;
            setInputValue(newValue);
            action('onChange')(newValue);
          }}
          value={inputValue}
          placeholder={text('placeholder', 'Place Holder')}
          onClear={() => {
            setInputValue('');
            action('onClear')();
          }}
          iconPrefix={getIconPath(select('iconPrefix', options, options.mdiComment))}
          isMultiline={isMultiline}
          errorMessage={isError}
        />
      );
    },
    { design },
  );
