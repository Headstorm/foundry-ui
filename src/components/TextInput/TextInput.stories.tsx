import React, { useState, useCallback } from 'react';
import { Story, Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import * as IconPaths from '@mdi/js';

import TextInput, { TextInputProps } from './TextInput';
import variants from '../../enums/variants';

const iconOptions = {
  none: '',
  ...IconPaths,
};

const maxLengthOptions = {
  5: 5,
  20: 20,
  100: 100,
  none: undefined,
};

const callbacks: {
  [i: string]: (event: React.ChangeEvent<HTMLTextAreaElement> & React.KeyboardEvent) => void;
} = {
  onChange: event => action('onChange')(event.target.value),
  debouncedOnChange: event => action('debouncedOnChange')(event.target.value),
  onClear: () => action('onClear')(),
  onFocus: () => action('onFocus')(),
  onBlur: () => action('onBlur')(),
  onInput: event => action('onInput')(event.target.value),
  onKeyPress: event => action('onKeyPress')(event.key),
  onKeyUp: event => action('onKeyUp')(event.key),
  onKeyDown: event => action('onKeyDown')(event.key),
};

type BasicTextInputProps = TextInputProps & {
  clearable: boolean;
};

export const BasicTextInput: Story<BasicTextInputProps> = ({
  clearable,
  onChange,
  onClear,
  ...args
}: BasicTextInputProps) => {
  const [inputValue, setInputValue] = useState('');

  const onChangeCallback = useCallback(
    event => {
      setInputValue(event.target.value);
      if (onChange) onChange(event);
    },
    [onChange],
  );
  const onClearCallback = useCallback(
    event => {
      setInputValue('');
      if (onClear) onClear(event);
    },
    [onClear],
  );

  return (
    <TextInput
      {...args}
      value={inputValue}
      onChange={onChangeCallback}
      onClear={clearable ? onClearCallback : undefined}
    />
  );
};
BasicTextInput.args = {
  'aria-label': 'textInput',
  debounceInterval: 150,
  disabled: false,
  placeholder: 'Placeholder',
  iconPrefix: 'none',
  isMultiline: false,
  rows: 0,
  cols: 0,
  isValid: true,
  errorMessage: '',
  variant: variants.outline,
  type: '',
  multiLineIsResizable: false,
  showCharacterCount: true,
  maxLength: 20,
  allowTextBeyondMaxLength: false,
  clearable: false,
  ...callbacks,
};

export const UncontrolledTextInput: Story<TextInputProps> = (args: TextInputProps) => {
  return <TextInput {...args} onClear={undefined} />;
};
UncontrolledTextInput.args = { ...BasicTextInput.args };

export default {
  title: 'TextInput',
  argTypes: {
    iconPrefix: {
      options: Object.keys(iconOptions),
      mapping: iconOptions,
      control: {
        type: 'select',
      },
    },
    maxLength: {
      options: Object.keys(maxLengthOptions),
      mapping: maxLengthOptions,
      control: {
        type: 'select',
      },
    },
    variant: {
      options: Object.keys(variants),
      mapping: variants,
      control: {
        type: 'radio',
      },
    },
  },
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/3r2G00brulOwr9j7F6JF59/Generic-UI-Style?node-id=102%3A29',
    },
  },
} as Meta;
