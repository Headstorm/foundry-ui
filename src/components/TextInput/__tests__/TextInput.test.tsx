import React from 'react';
import styled from 'styled-components';
import { render, configure, waitFor } from '@testing-library/react';
import TextInput from '../TextInput';
import { axe, toHaveNoViolations } from 'jest-axe';

expect.extend(toHaveNoViolations);
configure({ testIdAttribute: 'data-test-id' });

const testId = 'foundry-text-input';
const containerProps = { 'data-test-id': testId };

describe('TextInput', () => {
  describe('Accessibility Tests', () => {
    it('Should pass accessibility test with default props', async () => {
      const component = <TextInput aria-label="text-input"></TextInput>;
      const { container } = render(component);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
  describe('Ref tests', () => {
    it('containerRef.current should exist', async () => {
      const ref = React.createRef<HTMLDivElement>();
      const { getByTestId } = render(
        <TextInput containerRef={ref} containerProps={containerProps} />,
      );
      await waitFor(() => getByTestId(testId));
      expect(ref.current instanceof HTMLDivElement).toBeTruthy();
    });
    it('inputRef.current should exist', async () => {
      const ref = React.createRef<HTMLInputElement>();
      const { getByTestId } = render(<TextInput inputRef={ref} containerProps={containerProps} />);
      await waitFor(() => getByTestId(testId));
      expect(ref.current instanceof HTMLInputElement).toBeTruthy();
    });
    it('errorContainerRef.current should exist', async () => {
      const ref = React.createRef<HTMLDivElement>();
      const { getByTestId } = render(
        <TextInput
          errorContainerRef={ref}
          isValid={false}
          errorMessage={'error!'}
          containerProps={containerProps}
        />,
      );
      await waitFor(() => getByTestId(testId));
      expect(ref.current instanceof HTMLDivElement).toBeTruthy();
    });      
    it('characterCountRef.current should exist', async () => {
      const ref = React.createRef<HTMLDivElement>();
      const { getByTestId } = render(
        <TextInput
        showCharacterCount
        maxLength={20}
        characterCountRef={ref}
        containerProps={containerProps}
        />,
      );
      await waitFor(() => getByTestId(testId));
      expect(ref.current instanceof HTMLDivElement).toBeTruthy();
    });
  });
});
