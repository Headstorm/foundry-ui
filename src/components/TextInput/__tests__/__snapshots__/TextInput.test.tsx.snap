// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Rendering tests Shows TextInput with default props 1`] = `
.c0 {
  min-width: 10rem;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  border-radius: 0.25em;
  border: 1px solid #51575D;
  background-color: #fff;
}

.c1 {
  border: 0 none;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  outline: 0 none;
  font-size: 1em;
  padding: 0.5rem;
  background-color: transparent;
}

.c1:focus {
  outline: none;
  box-shadow: 0 0 5px 0.150rem #64ACC5;
}

<div>
  <div
    class=" c0"
    data-test-id="foundry-text-input"
  >
    <input
      class=" c1"
      cols="10"
      rows="10"
      type="text"
    />
  </div>
</div>
`;

exports[`TextInput Rendering tests Shows TextInput with fill variant 1`] = `
.c0 {
  min-width: 10rem;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  border-radius: 0.25em;
  border: 1px solid transparent;
  border-bottom: 1px solid #51575D;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  background-color: #e6e6e6;
}

.c1 {
  border: 0 none;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  outline: 0 none;
  font-size: 1em;
  padding: 0.5rem;
  background-color: transparent;
}

.c1:focus {
  outline: none;
  box-shadow: 0 0 5px 0.150rem #64ACC5;
}

<div>
  <div
    class=" c0"
    data-test-id="foundry-text-input"
  >
    <input
      class=" c1"
      cols="10"
      rows="10"
      type="text"
    />
  </div>
</div>
`;

exports[`TextInput Rendering tests Shows TextInput with text variant 1`] = `
.c0 {
  min-width: 10rem;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  border-radius: 0.25em;
  border: 1px solid transparent;
  background-color: #fff;
}

.c1 {
  border: 0 none;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  outline: 0 none;
  font-size: 1em;
  padding: 0.5rem;
  background-color: transparent;
}

.c1:focus {
  outline: none;
  box-shadow: 0 0 5px 0.150rem #64ACC5;
}

<div>
  <div
    class=" c0"
    data-test-id="foundry-text-input"
  >
    <input
      class=" c1"
      cols="10"
      rows="10"
      type="text"
    />
  </div>
</div>
`;
